cache/surrogate/surrogate.go

```diff
@@ -0,0 +1,97 @@
+package surrogate
+
+// The Surrogate key system is like the ykey one in Varnish and the same as Fastly.
+// You can send a PURGE request on the Souin API endpoint with the Surrogate header to invalidate the filled keys.
+// The Surrogate header CAN contains one or multiple keys separated by a comma as mentioned in the RFC.
```

Comment:
IMHO we could add a directive to soft delete like `abc;soft` or `abc;keep` to the `Surrogate-Control`

Reply:
I don't think you should deviate from `Surrogate-Control` usage in the wild so much. It may not be a standardized header officially, but it probably would confuse other users who have experience with the header from other solutions where it is treated as a targeted form of `Cache-Control`.

Fastly uses `grace` and `keep` under the hood due to Varnish heritage. [They explain it well here](https://docs.fastly.com/en/guides/serving-stale-content#interaction-with-grace), where `grace` is equivalent to the `Cache-Control` directive `stale-if-error` and `keep` is equivalent to how long Fastly keeps the stale cache without returning it, if a request arrives and the backend responds with a 304 the resource does not need to be transferred to Souin again, and the stale entry can be marked as fresh once more to serve.

There is a [related Fastly doc on staleness](https://developer.fastly.com/learning/concepts/stale/) with a visualization of the freshness => stale timeline where `stale-while-revalidate`/`stale-if-error` and expired status apply, along with a matrix table of Content + Origin state for conveying the rules of how cached responses are handled.

A soft purge with Fastly will make the cache entry stale, so `stale-while-revalidate` applies, followed by `stale-if-error`, then expired (`Etag`/`Last-Modified` check while not evicted). Fastly allows you to soft purge a URL resource or Surrogate Key (via endpoint) with the `Fastly-Soft-Purge: 1` header ([soft purge docs](https://docs.fastly.com/en/guides/soft-purges)).

Reply:
So, it could be built on top of the Cache-Control header then and send the directive in both `Surrogate-Control` and `Cache-Control`.

The keys could be like `abc;stale-{any stale type}` to soft purge, WDYT ?

Reply:
No.

I understand that I have written quite a bit of information for you to consider. But I'm not sure if you're following or understanding much of it?

---

There is no reason or need for what you're proposing. What is wrong with the existing approaches to purging that you consider are bad and need a different approach?

**Soft purge** makes a cached resource no longer fresh, it enters stale state. That is all that needs to happen, there is no reason for you to want to specifically skip `stale-while-revalidate` to `stale-if-error` or the expired state that is kept.

**Hard purge** removes the resource from cache proxy completely. Both purge types are desirable.

**`Surrogate-Control`** is already well established but never became a standard. `CDN-Cache-Control` (TCC) is becoming one. Neither are headers that you'd be able to officially get `Surrogate-Key` into, they are intended to be `Cache-Control` directive without scoping to surrogate keys.

Reply:
> What is wrong with the existing approaches to purging that you consider are bad and need a different approach? 
https://github.com/darkweak/souin/pull/134#discussion_r706702860
If you absolutely want to keep the existing approches (such as the ones defined from Varnish, CF, Faslty, ...), why do I continue to develop this cache system? Why not trying to make something newer and more consistent?

Reply:
> why should I continue to develop this cache system?

You're building a great cache system that is versatile, configurable and easy to use AFAIK. In addition you're providing features such as Surrogate Keys with purging, features that each of these services typically place behind a pay wall IIRC.

That is unrelated to standardization/specification. You're still able to provide the above regardless of how you implement `Surrogate-Key` and `Surrogate-Control`/`CDN-Cache-Control` headers and related functionality.

I do strongly advise you assess what the industry leaders are doing, and only deviate from that when you have a strong justification to do so, not due to NIH syndrome.

Presently, I have not seen any compelling reasons for the approach you propose over what is already out in the wild. I have asked for practical examples, but have yet to receive any. If believe there is valid value in a different approach, please convey that value with examples. Earlier I linked to two scenarios from Fastly, anything like that, or why their approach is problematic would be appreciated.

---

> Why not trying to make something newer and more consistent?

![NIH syndrome](https://imgs.xkcd.com/comics/standards.png)

Please don't contribute to the problem just to be different.

`Surrogate-Control` is roughly the same across providers but not standardized successfully. That differs with `CDN-Cache-Control` (where CDN can be other targets eg Souin), which Akamai, Cloudflare and Fastly all collaborated together on. This works as I have described earlier, I very much doubt you'll have success proposing a change to that or `Cache-Control`, especially if you lack any compelling reasons to do so.

`Surrogate-Key`/`Cache-Tag` on the other hand isn't standardized. For same reasons `Surrogate-Control` did not get standardized, I think you'll find the same problem with `Surrogate-Key`. You're more than welcome to try, but even if it's not an IETF official standard, creating a solid specification document is good for the community to adopt.

My main red flag right now is that you are choosing to ignore existing successful implementations of `Surrogate-Key` and re-define it yourself when there doesn't appear to be a need? What is wrong with the approaches from Fastly or Cloudflare? Their header values provide the associated surrogate keys to a cache resource, that's all that is required?

Reply:
Okay, I'll be compliant with the Akamai/CF/Fastly implementation and drop the RFC redaction.

Reply:
You were right from the start, I should not try to reinvent the wheel on this but I didn't want to hear it. 😄

Reply:
> I should not try to reinvent the wheel on this but I didn't want to hear it.

It is ok! I think it is an easy trap to fall into, I know I have done it myself a few times :sweat_smile: 

I hope I have not come off too harsh with my feedback at times. I really want to see Souin do well!